### PR TITLE
feat: add Helm chart to install Tracee with Postee

### PR DESCRIPTION
**Pull request by danielpacak**
## Initial Checklist

- [ ] There is an issue describing the need for this PR.
- [x] Git log contains summary of the change.
- [x] Git log contains motivation and context of the change.
- [ ] If part of an EPIC, PR git log contains EPIC number.
- [ ] If part of an EPIC, PR was added to EPIC description.

## Description (git log)

This patch adds basic Helm chart templates to install Tracee
as a DaemonSet in a Kubernetes cluster.

You can install tracee Helm release with the following command:

    git clone https://github.com/aquasecurity/tracee.git
    cd tracee
    helm repo add aqua-charts https://aquasecurity.github.io/helm-charts
    helm dependency update ./deploy/helm/tracee
    helm install tracee ./deploy/helm/tracee \
      --namespace tracee-system --create-namespace

When you want to test the template rendering, but not actually
installing anything, you can use:

    helm install --debug --dry-run tracee ./deploy/helm/tracee \
      --namespace tracee-system --create-namespace

By default, Tracee is installed with Postee. You can disable Postee
with the `postee.enabled=false` value:

    helm install tracee ./deploy/helm/tracee \
      --namespace tracee-system --create-namespace \
      --set postee.enabled=false

To see the entire generated YAML run the following command:

    helm get manifest tracee --namespace tracee-system

To uninstall Tracee run:

    helm uninstall tracee -n tracee-system

## Type of change

- [ ] Bug fix (non-breaking change fixing an issue, preferable).
- [ ] Quick fix (minor non-breaking change requiring no issue, use with care)
- [ ] Code refactor (code improvement and/or code removal)
- [x] New feature (non-breaking change adding functionality).
- [ ] Breaking change (cause existing functionality not to work as expected).

## How Has This Been Tested?

Manual test with helm install command. See commit message for test scenario.

## Final Checklist:

Pick "Bug Fix" or "Feature", delete the other and mark appropriate checks.

- [x] I have made corresponding changes to the documentation.
- [x] My code follows the style guidelines (C and Go) of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented all functions/methods created explaining what they do.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] My changes generate no new warnings.
- [ ] I have added tests that prove my fix, or feature, is effective.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published before.

## Git Log Checklist:

My commits logs have:

- [x] Subject starts with "subsystem|file: description".
- [x] Do not end the subject line with a period.
- [x] Limit the subject line to 50 characters.
- [x] Separate subject from body with a blank line.
- [x] Use the imperative mood in the subject line.
- [x] Wrap the body at 72 characters.
- [x] Use the body to explain what and why instead of how.

### CODE REVIEW

File: deploy/helm/tracee/templates/NOTES.txt [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+{{- if .Values.postee.enabled }}
+
+Detections are also sent to the webhook endpoint exposed by Postee:
+
+$ kubectl logs -f statefulset/{{ include "tracee.fullname" . }}-postee -n {{ .Release.Namespace }}
+
+You can integrate Postee with email server, Slack channel, and any other supported
+data sink to forward threat alerts for forensic auditing. For more details on Postee,
+see its documentation on https://aquasecurity.github.io/postee.
```
Comment (simar7) [inline deploy/helm/tracee/templates/NOTES.txt:23]: Just a nit:
```suggestion
see its documentation on https://aquasecurity.github.io/postee/latest.
```

**Comment by simar7** [comment]
So I was trying to play around with it to try it out as I want to learn more about helm and ran into this:
```
$ helm dependency build ./deploy/helm/tracee
WARNING: Kubernetes configuration file is group-readable. This is insecure. Location: /home/vagrant/.kube/config
WARNING: Kubernetes configuration file is world-readable. This is insecure. Location: /home/vagrant/.kube/config
Error: the lock file (Chart.lock) is out of sync with the dependencies file (Chart.yaml). Please update the dependencies
```

Running a `helm dep update` led me to this:
```
diff --git a/deploy/helm/tracee/Chart.lock b/deploy/helm/tracee/Chart.lock
index 24ac3a9..fb83c69 100644
--- a/deploy/helm/tracee/Chart.lock
+++ b/deploy/helm/tracee/Chart.lock
@@ -2,5 +2,5 @@ dependencies:
 - name: postee
   repository: https://aquasecurity.github.io/helm-charts
   version: v2.6.0
-digest: sha256:6b5654cf403b9b1423ae9b40d885c7514beba526a45807204005441fae574e14
-generated: "2022-06-08T22:16:16.71+02:00"
+digest: sha256:840d3ac8afb52fc0cc060cf21be26f3026069d124ecda255a4c0114abbf156c8
+generated: "2022-06-10T00:29:38.653103402Z"

```

After this Tracee seems to run fine but Postee fails to get ready. Here's some debug info:
```
vagrant@ubuntu-impish:/repos/tracee$ k get pods --namespace=tracee-system
NAME                               READY   STATUS    RESTARTS   AGE
tracee-postee-0                    0/1     Pending   0          104s
tracee-posteeui-5cd54bd848-fchtv   0/1     Pending   0          104s
tracee-wcsmk                       1/1     Running   0          104s
vagrant@ubuntu-impish:/repos/tracee$ k describe pod tracee-postee-0     --namespace=tracee-system
Name:           tracee-postee-0
Namespace:      tracee-system
Priority:       0
Node:           <none>
Labels:         app.kubernetes.io/instance=tracee
                app.kubernetes.io/name=postee
                controller-revision-hash=tracee-postee-55b9d8d4d6
                statefulset.kubernetes.io/pod-name=tracee-postee-0
Annotations:    <none>
Status:         Pending
IP:             
IPs:            <none>
Controlled By:  StatefulSet/tracee-postee
Init Containers:
  setting-db:
    Image:      busybox:1.34
    Port:       <none>
    Host Port:  <none>
    Command:
      /bin/chown
      -R
      1099
      /server/database
    Environment:  <none>
    Mounts:
      /server/database from tracee-postee-db (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-jznl7 (ro)
  setting-cfg:
    Image:      busybox:1.34
    Port:       <none>
    Host Port:  <none>
    Command:
      /bin/cp
      /k8s/cfg.yaml
      /data/cfg.yaml
    Environment:  <none>
    Mounts:
      /data from tracee-postee-config (rw)
      /k8s from tracee-postee-configmap-vol (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-jznl7 (ro)
Containers:
  postee:
    Image:       aquasec/postee:latest
    Ports:       8082/TCP, 8445/TCP
    Host Ports:  0/TCP, 0/TCP
    Limits:
      cpu:     500m
      memory:  256Mi
    Requests:
      cpu:     200m
      memory:  128Mi
    Environment:
      POSTEE_CFG:    /data/cfg.yaml
      POSTEE_DEBUG:  not
    Mounts:
      /data from tracee-postee-config (rw)
      /server/database from tracee-postee-db (rw)
      /server/rego-filters/custom from tracee-postee-filters (rw)
      /server/rego-templates/custom from tracee-postee-rego-template (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-jznl7 (ro)
Conditions:
  Type           Status
  PodScheduled   False 
Volumes:
  tracee-postee-filters:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  tracee-postee-filters-tracee-postee-0
    ReadOnly:   false
  tracee-postee-db:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  tracee-postee-db-tracee-postee-0
    ReadOnly:   false
  tracee-postee-config:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  tracee-postee-config-tracee-postee-0
    ReadOnly:   false
  tracee-postee-rego-template:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  tracee-postee-rego-template-tracee-postee-0
    ReadOnly:   false
  tracee-postee-configmap-vol:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      tracee-postee-cfg
    Optional:  false
  kube-api-access-jznl7:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   Burstable
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason            Age                 From               Message
  ----     ------            ----                ----               -------
  Warning  FailedScheduling  2m8s                default-scheduler  0/1 nodes are available: 1 pod has unbound immediate PersistentVolumeClaims.
  Warning  FailedScheduling  58s (x1 over 2m7s)  default-scheduler  0/1 nodes are available: 1 pod has unbound immediate PersistentVolumeClaims.
```

Any ideas what could be missing?

**Comment by danielpacak** [comment]
> Any ideas what could be missing?

I think you got a good indication that something is wrong with unbound (or *Pending*) PersistentVolumeClaims that are use by Postee Helm subchart:

```
Events:
  Type     Reason            Age                 From               Message
  ----     ------            ----                ----               -------
  Warning  FailedScheduling  2m8s                default-scheduler  0/1 nodes are available: 1 pod has unbound immediate PersistentVolumeClaims.
  Warning  FailedScheduling  58s (x1 over 2m7s)  default-scheduler  0/1 nodes are available: 1 pod has unbound immediate PersistentVolumeClaims.
```

One reason for that is undefined StorageClass or PersistentVolume. In Vagrant vm and MicroK8s this can be done by enabling the *hostpath-storage* add-on. If the hostpath provisioner is up and running you should see all PVCs in *Bound* status (not *Pending*):

```console
$ kubectl get pvc -n tracee-system
NAME                                          STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS        AGE
tracee-postee-config-tracee-postee-0          Bound    pvc-bcdcebe2-d0d1-4bd5-ba5a-723a2001db80   100Mi      RWO            microk8s-hostpath   3m24s
tracee-postee-filters-tracee-postee-0         Bound    pvc-25262cae-7197-496f-9e9a-89ff477f0227   100Mi      RWO            microk8s-hostpath   3m24s
tracee-postee-rego-template-tracee-postee-0   Bound    pvc-62ac8e89-0a85-4c0d-9f97-1a37c21b0161   100Mi      RWO            microk8s-hostpath   3m24s
tracee-postee-db-tracee-postee-0              Bound    pvc-68dd8f4b-c0ec-4024-aad3-6d328080faec   1Gi        RWO            microk8s-hostpath   3m24s
```

I noticed that another problem you'll bump into with Vagrant vm will be Tracee pod not being able to resolve the webhook endpoint URL (e.g., http://tracee-postee:8080). This is because the DNS service is not enabled in MicroK8s by default. Therefore, I've updated *Vagrantfile* to enable both *hostpath-storage* and *dns* add-ons.

**Comment by simar7** [comment]
> > Any ideas what could be missing?
> 
> Most likely it's because of unbound (or _Pending_) PersistentVolumeClaims that are use by Postee Helm subchart:
> 
> ```
> Events:
>   Type     Reason            Age                 From               Message
>   ----     ------            ----                ----               -------
>   Warning  FailedScheduling  2m8s                default-scheduler  0/1 nodes are available: 1 pod has unbound immediate PersistentVolumeClaims.
>   Warning  FailedScheduling  58s (x1 over 2m7s)  default-scheduler  0/1 nodes are available: 1 pod has unbound immediate PersistentVolumeClaims.
> ```
> 
> One reason for that is undefined StorageClass or PersistentVolume. In Vagrant vm and MicroK8s this can be done by enabling the _hostpath-storage_ add-on. If the hostpath provisioner is up and running you should see all PVCs in _Bound_ status (not _Pending_):
> 
> ```
> $ kubectl get pvc -n tracee-system
> NAME                                          STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS        AGE
> tracee-postee-config-tracee-postee-0          Bound    pvc-bcdcebe2-d0d1-4bd5-ba5a-723a2001db80   100Mi      RWO            microk8s-hostpath   3m24s
> tracee-postee-filters-tracee-postee-0         Bound    pvc-25262cae-7197-496f-9e9a-89ff477f0227   100Mi      RWO            microk8s-hostpath   3m24s
> tracee-postee-rego-template-tracee-postee-0   Bound    pvc-62ac8e89-0a85-4c0d-9f97-1a37c21b0161   100Mi      RWO            microk8s-hostpath   3m24s
> tracee-postee-db-tracee-postee-0              Bound    pvc-68dd8f4b-c0ec-4024-aad3-6d328080faec   1Gi        RWO            microk8s-hostpath   3m24s
> ```
> 
> I noticed that another problem you'll bump into with Vagrant vm will be Tracee pod not being able to resolve the webhook endpoint URL (e.g., http://tracee-postee:8080). This is because the DNS service is not enabled in MicroK8s by default. Therefore, I've updated _Vagrantfile_ to enable both _hostpath-storage_ and _dns_ add-ons.

Bingo, looks like that did the trick!

**Review by simar7 — APPROVED** [review]
LGTM! Just a small nit about the Vagrant port btw: Looks like recently docker has made some updates to Docker Desktop and now it also listens on port 8000:

```
➜  tracee-aqua git:(helm) ✗ sudo lsof -i -P | grep LISTEN | grep 8000
Password:
com.docke 37822          simar  207u  IPv6 0xb95fdb335448033f      0t0    TCP *:8000 (LISTEN)
➜  tracee-aqua git:(helm) ✗ sudo lsof -i -P | grep LISTEN | grep 8000
```

This causes a Vagrant box provision conflict. Maybe we can change this to another port?

**Comment by rafaeldtinoco** [comment]
@simar7 let me know once you're good for us to merge this one. Thanks

**Comment by danielpacak** [comment]
> LGTM! Just a small nit about the Vagrant port btw: Looks like recently docker has made some updates to Docker Desktop and now it also listens on port 8000:
> 
> ```
> ➜  tracee-aqua git:(helm) ✗ sudo lsof -i -P | grep LISTEN | grep 8000
> Password:
> com.docke 37822          simar  207u  IPv6 0xb95fdb335448033f      0t0    TCP *:8000 (LISTEN)
> ```
> 
> This causes a Vagrant box provision conflict. Maybe we can change this to another port?

I'm running the latest Docker Desktop on macOS and by default it does not use port 8000. Isn't it a container or extension running in your environment that conflicts with port 8000 forwarded by Vagrant / Hypervisor? In any case a dedicated patch to Vagrant file that uses another port number is a good idea. I would keep it out of scope of this PR.

**Comment by rafaeldtinoco** [comment]
Okay, based on latest feedback from @danielpacak I'm going ahead and merging this, we can always fix it later if needed, but seems like we're good shape for now. Thanks!

**Pull request merged**